ccl/streamingccl/streamingest: TestStreamIngestionProcessor failed

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3286193&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3286193&tab=artifacts#/) on master @ [90809c048d05f923a67ce9b89597b2779fc73e32](https://github.com/cockroachdb/cockroach/commits/90809c048d05f923a67ce9b89597b2779fc73e32):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor987361220
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 26265 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0x4b19f00, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x21c
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc00897da10, 0xc005e60fc0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
        Leaked goroutine: goroutine 26266 [semacquire]:
        sync.runtime_Semacquire(0xc00897da20)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc00897da18)
        	/usr/local/go/src/sync/waitgroup.go:130 +0x65
        golang.org/x/sync/errgroup.(*Group).Wait(0xc00897da10, 0xc00897c158, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x31
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc00897da10, 0xc0022f7100, 0xc0026957a0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x3c
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x314
--- FAIL: TestStreamIngestionProcessor (6.72s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3286194&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3286194&tab=artifacts#/) on master @ [90809c048d05f923a67ce9b89597b2779fc73e32](https://github.com/cockroachdb/cockroach/commits/90809c048d05f923a67ce9b89597b2779fc73e32):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor094532175
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 14077 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc001a2af50, 0xc001a2ae38)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc002eb44e0, 0xc005e21640)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 14078 [semacquire]:
        sync.runtime_Semacquire(0xc002eb44f0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc002eb44e8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc002eb44e0, 0xc0019e4c00, 0xc002e53680)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc002eb44e0, 0xc00519dc00, 0xc000b6c5a0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (17.66s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3291107&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3291107&tab=artifacts#/) on master @ [e723a85739a2e1b514f5de3221a2177a3a37d0e6](https://github.com/cockroachdb/cockroach/commits/e723a85739a2e1b514f5de3221a2177a3a37d0e6):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor734325929
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 26135 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0x0, 0x100010000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x21c
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0037ff530, 0xc003c017c0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
        Leaked goroutine: goroutine 26136 [semacquire]:
        sync.runtime_Semacquire(0xc0037ff540)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0037ff538)
        	/usr/local/go/src/sync/waitgroup.go:130 +0x65
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0037ff530, 0x0, 0x4b4fa70)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x31
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0037ff530, 0xc002e68300, 0xc002c5c6c0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x3c
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x314
--- FAIL: TestStreamIngestionProcessor (6.90s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3291108&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3291108&tab=artifacts#/) on master @ [e723a85739a2e1b514f5de3221a2177a3a37d0e6](https://github.com/cockroachdb/cockroach/commits/e723a85739a2e1b514f5de3221a2177a3a37d0e6):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor785303510
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 13893 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc002588f50, 0xc002588e38)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc003bc5410, 0xc003bcda00)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 13894 [semacquire]:
        sync.runtime_Semacquire(0xc003bc5420)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc003bc5418)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc003bc5410, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc003bc5410, 0xc001ad5c00, 0xc00267b6e0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (17.08s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3293952&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3293952&tab=artifacts#/) on master @ [812ba837ed23a3a245faecd59700d82297abb4d4](https://github.com/cockroachdb/cockroach/commits/812ba837ed23a3a245faecd59700d82297abb4d4):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor784128178
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 23319 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0x0, 0x4b58e18)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x21c
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc007972900, 0xc0006b7c40)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
        Leaked goroutine: goroutine 23320 [semacquire]:
        sync.runtime_Semacquire(0xc007972910)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc007972908)
        	/usr/local/go/src/sync/waitgroup.go:130 +0x65
        golang.org/x/sync/errgroup.(*Group).Wait(0xc007972900, 0x0, 0x4b58e18)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x31
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc007972900, 0xc007b95100, 0xc0037190e0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x3c
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x314
--- FAIL: TestStreamIngestionProcessor (9.40s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3294717&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3294717&tab=artifacts#/) on master @ [9a0286adc7db55c414ac0b0b0c54ef4cbc138ba0](https://github.com/cockroachdb/cockroach/commits/9a0286adc7db55c414ac0b0b0c54ef4cbc138ba0):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor686050535
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7232 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc0027e7750, 0x2)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0030298c0, 0xc0034b3d80)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7233 [semacquire]:
        sync.runtime_Semacquire(0xc0030298d0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0030298c8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0030298c0, 0x18, 0xc0027eb638)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0030298c0, 0xc00577dc00, 0xc00579d1a0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (18.59s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3296422&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3296422&tab=artifacts#/) on master @ [fdf147ce8aee7fb7e59f6285850a5bc59c19fef3](https://github.com/cockroachdb/cockroach/commits/fdf147ce8aee7fb7e59f6285850a5bc59c19fef3):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor425563021
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7235 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc003dc3f50, 0xc003dc3e38)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0039c8ba0, 0xc005aac700)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7236 [semacquire]:
        sync.runtime_Semacquire(0xc0039c8bb0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0039c8ba8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0039c8ba0, 0x18, 0xc003dc2638)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0039c8ba0, 0xc002d85800, 0xc003d499e0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (18.70s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3296884&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3296884&tab=artifacts#/) on master @ [3928f1963833fbf51ae47bd2a42ae6a200ebbb14](https://github.com/cockroachdb/cockroach/commits/3928f1963833fbf51ae47bd2a42ae6a200ebbb14):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor467501966
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7156 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc00338cf50, 0xc00338ce38)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc003bf28a0, 0xc001e0a880)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7157 [semacquire]:
        sync.runtime_Semacquire(0xc003bf28b0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc003bf28a8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc003bf28a0, 0x18, 0xc00338fe38)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc003bf28a0, 0xc007075500, 0xc00359fb00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (18.08s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3296883&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3296883&tab=artifacts#/) on master @ [3928f1963833fbf51ae47bd2a42ae6a200ebbb14](https://github.com/cockroachdb/cockroach/commits/3928f1963833fbf51ae47bd2a42ae6a200ebbb14):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor479980974
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 21158 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc0003c82a0, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x21c
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0050a0810, 0xc001406200)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
        Leaked goroutine: goroutine 21159 [semacquire]:
        sync.runtime_Semacquire(0xc0050a0820)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0050a0818)
        	/usr/local/go/src/sync/waitgroup.go:130 +0x65
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0050a0810, 0x0, 0x4b700c0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x31
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0050a0810, 0xc004c63500, 0xc0014f8a80)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x3c
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x314
--- FAIL: TestStreamIngestionProcessor (7.18s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3301665&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3301665&tab=artifacts#/) on master @ [0880e83e30ee5eb9aab7bb2297324e098d028225](https://github.com/cockroachdb/cockroach/commits/0880e83e30ee5eb9aab7bb2297324e098d028225):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor075799508
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 20481 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0x0, 0x1ad)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x21c
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc001ca5b90, 0xc004b72a80)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
        Leaked goroutine: goroutine 20546 [semacquire]:
        sync.runtime_Semacquire(0xc001ca5ba0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc001ca5b98)
        	/usr/local/go/src/sync/waitgroup.go:130 +0x65
        golang.org/x/sync/errgroup.(*Group).Wait(0xc001ca5b90, 0x0, 0x4b6ea90)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x31
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc001ca5b90, 0xc001f08a00, 0xc004c80240)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x3c
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x314
--- FAIL: TestStreamIngestionProcessor (7.03s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3301666&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3301666&tab=artifacts#/) on master @ [0880e83e30ee5eb9aab7bb2297324e098d028225](https://github.com/cockroachdb/cockroach/commits/0880e83e30ee5eb9aab7bb2297324e098d028225):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor126631159
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 6835 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc001c8af50, 0xc002c91bb0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc003e4a8a0, 0xc00271b780)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 6836 [semacquire]:
        sync.runtime_Semacquire(0xc003e4a8b0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc003e4a8a8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc003e4a8a0, 0x59d360c, 0xc00689ce00)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc003e4a8a0, 0xc00689ce00, 0xc00012c240)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (16.87s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3302974&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3302974&tab=artifacts#/) on master @ [5dc07daf763fd59d390a4089ab517abcda875467](https://github.com/cockroachdb/cockroach/commits/5dc07daf763fd59d390a4089ab517abcda875467):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor519130782
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7455 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc000fd3750, 0xc000fd3638)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0005d62d0, 0xc001dd41c0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7456 [semacquire]:
        sync.runtime_Semacquire(0xc0005d62e0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0005d62d8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0005d62d0, 0x18, 0xc002bab638)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0005d62d0, 0xc005762300, 0xc0051afbc0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (20.61s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3303050&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3303050&tab=artifacts#/) on master @ [ee3efd6b1e24a3e1676778f5028fa0a35266f683](https://github.com/cockroachdb/cockroach/commits/ee3efd6b1e24a3e1676778f5028fa0a35266f683):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor159357265
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7065 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc00306bf50, 0xc00306be38)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc003589ad0, 0xc002c8f840)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7066 [semacquire]:
        sync.runtime_Semacquire(0xc003589ae0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc003589ad8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc003589ad0, 0xc0001a6010, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc003589ad0, 0xc007f4d500, 0xc00642c7e0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (18.08s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3303444&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3303444&tab=artifacts#/) on master @ [ee3efd6b1e24a3e1676778f5028fa0a35266f683](https://github.com/cockroachdb/cockroach/commits/ee3efd6b1e24a3e1676778f5028fa0a35266f683):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor929439916
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 20516 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc000122240, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x21c
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0028142a0, 0xc00361df00)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
        Leaked goroutine: goroutine 20517 [semacquire]:
        sync.runtime_Semacquire(0xc0028142b0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0028142a8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0x65
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0028142a0, 0x0, 0x4baae10)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x31
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0028142a0, 0xc001551500, 0xc0026f4840)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x3c
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x314
--- FAIL: TestStreamIngestionProcessor (6.94s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3303445&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3303445&tab=artifacts#/) on master @ [ee3efd6b1e24a3e1676778f5028fa0a35266f683](https://github.com/cockroachdb/cockroach/commits/ee3efd6b1e24a3e1676778f5028fa0a35266f683):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor516815942
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 6782 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc003669f50, 0xc00195f100)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0032a08d0, 0xc001ded280)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 6783 [semacquire]:
        sync.runtime_Semacquire(0xc0032a08e0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0032a08d8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0032a08d0, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0032a08d0, 0xc00195f100, 0xc00041fa40)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (16.38s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3305264&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3305264&tab=artifacts#/) on master @ [04a41e7915f4a89dcc1d0dbd92466c6adf79ec9f](https://github.com/cockroachdb/cockroach/commits/04a41e7915f4a89dcc1d0dbd92466c6adf79ec9f):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor678164478
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 20607 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0x4bafc10, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x21c
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc002433aa0, 0xc005bffd40)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x59
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x66
        Leaked goroutine: goroutine 20608 [semacquire]:
        sync.runtime_Semacquire(0xc002433ab0)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc002433aa8)
        	/usr/local/go/src/sync/waitgroup.go:130 +0x65
        golang.org/x/sync/errgroup.(*Group).Wait(0xc002433aa0, 0xc005c4ff58, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x31
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc002433aa0, 0xc0060ce300, 0xc001fa8c00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x3c
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x314
--- FAIL: TestStreamIngestionProcessor (7.03s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3305265&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3305265&tab=artifacts#/) on master @ [04a41e7915f4a89dcc1d0dbd92466c6adf79ec9f](https://github.com/cockroachdb/cockroach/commits/04a41e7915f4a89dcc1d0dbd92466c6adf79ec9f):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor480714092
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7098 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc003a83750, 0xc003a83638)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0035e0060, 0xc00351f4c0)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7099 [semacquire]:
        sync.runtime_Semacquire(0xc0035e0070)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc0035e0068)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc0035e0060, 0x18, 0xc003a83e38)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc0035e0060, 0xc002584e00, 0xc003aba540)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (17.80s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3307148&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3307148&tab=artifacts#/) on master @ [becc3f9d1ab31a88c7980abdea495d817087ce06](https://github.com/cockroachdb/cockroach/commits/becc3f9d1ab31a88c7980abdea495d817087ce06):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor355980145
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7197 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc001b42750, 0xc001b42638)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc00550b230, 0xc003979940)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7198 [semacquire]:
        sync.runtime_Semacquire(0xc00550b240)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc00550b238)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc00550b230, 0x18, 0xc001b41638)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc00550b230, 0xc0028ec300, 0xc0046f6840)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (18.61s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3308538&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3308538&tab=artifacts#/) on master @ [39d59f1fc900f96a7256cd89fede44d9424356c4](https://github.com/cockroachdb/cockroach/commits/39d59f1fc900f96a7256cd89fede44d9424356c4):


```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStreamIngestionProcessor979896885
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:268: -- test log scope end --
    stream_ingestion_processor_test.go:268: Leaked goroutine: goroutine 7207 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func1(0xc0035a3f50, 0x968add8)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:376 +0x2b4
        golang.org/x/sync/errgroup.(*Group).Go.func1(0xc003c55080, 0xc003d60080)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x95
        created by golang.org/x/sync/errgroup.(*Group).Go
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x74
        Leaked goroutine: goroutine 7208 [semacquire]:
        sync.runtime_Semacquire(0xc003c55090)
        	/usr/local/go/src/runtime/sema.go:56 +0x45
        sync.(*WaitGroup).Wait(0xc003c55088)
        	/usr/local/go/src/sync/waitgroup.go:130 +0xe5
        golang.org/x/sync/errgroup.(*Group).Wait(0xc003c55080, 0xc002529080, 0xc00737c480)
        	/go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:40 +0x45
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge.func2(0xc003c55080, 0xc004c2ae00, 0xc00794c5a0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:390 +0x56
        created by github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).merge
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:389 +0x4ea
--- FAIL: TestStreamIngestionProcessor (18.87s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestStreamIngestionProcessor PKG=./pkg/ccl/streamingccl/streamingest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/bulk-io pbardea
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/base/testing_knobs.go
+++ b/pkg/base/testing_knobs.go
@@ -37,6 +37,7 @@ type TestingKnobs struct {
 	TenantTestingKnobs      ModuleTestingKnobs
 	JobsTestingKnobs        ModuleTestingKnobs
 	BackupRestore           ModuleTestingKnobs
+	Streaming               ModuleTestingKnobs
 	MigrationManager        ModuleTestingKnobs
 	IndexUsageStatsKnobs    ModuleTestingKnobs
 	SQLStatsKnobs           ModuleTestingKnobs

--- a/pkg/ccl/streamingccl/event.go
+++ b/pkg/ccl/streamingccl/event.go
@@ -116,3 +116,8 @@ func MakeKVEvent(kv roachpb.KeyValue) Event {
 func MakeCheckpointEvent(resolvedTimestamp hlc.Timestamp) Event {
 	return checkpointEvent{resolvedTimestamp: resolvedTimestamp}
 }
+
+// MakeGenerationEvent creates an GenerationEvent.
+func MakeGenerationEvent() Event {
+	return generationEvent{}
+}

--- a/pkg/ccl/streamingccl/streamclient/cockroach_sinkless_replication_client.go
+++ b/pkg/ccl/streamingccl/streamclient/cockroach_sinkless_replication_client.go
@@ -11,6 +11,7 @@ package streamclient
 import (
 	"context"
 	gosql "database/sql"
+	"database/sql/driver"
 	"fmt"
 	"strconv"
 
@@ -124,7 +125,15 @@ func (m *sinklessReplicationClient) ConsumePartition(
 			}
 		}
 		if err := rows.Err(); err != nil {
-			errCh <- err
+			if errors.Is(err, driver.ErrBadConn) {
+				select {
+				case eventCh <- streamingccl.MakeGenerationEvent():
+				case <-ctx.Done():
+					errCh <- ctx.Err()
+				}
+			} else {
+				errCh <- err
+			}
 			return
 		}
 	}()

--- a/pkg/ccl/streamingccl/streamclient/cockroach_sinkless_replication_client_test.go
+++ b/pkg/ccl/streamingccl/streamclient/cockroach_sinkless_replication_client_test.go
@@ -112,4 +112,17 @@ INSERT INTO d.t2 VALUES (2);
 		feed.ObserveResolved(ctx, secondObserved.Value.Timestamp)
 		cancelIngestion()
 	})
+
+	t.Run("stream-address-disconnects", func(t *testing.T) {
+		clientCtx, cancelIngestion := context.WithCancel(ctx)
+		eventCh, errCh, err := client.ConsumePartition(clientCtx, pa, startTime)
+		require.NoError(t, err)
+		feedSource := &channelFeedSource{eventCh: eventCh, errCh: errCh}
+		feed := streamingtest.MakeReplicationFeed(t, feedSource)
+
+		h.SysServer.Stopper().Stop(clientCtx)
+
+		require.True(t, feed.ObserveGeneration(clientCtx))
+		cancelIngestion()
+	})
 }

--- a/pkg/ccl/streamingccl/streamingest/BUILD.bazel
+++ b/pkg/ccl/streamingccl/streamingest/BUILD.bazel
@@ -85,6 +85,7 @@ go_test(
         "//pkg/security/securitytest",
         "//pkg/server",
         "//pkg/settings/cluster",
+        "//pkg/sql",
         "//pkg/sql/execinfra",
         "//pkg/sql/execinfrapb",
         "//pkg/sql/sem/tree",

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
@@ -21,6 +21,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/bulk"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
@@ -31,6 +32,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"golang.org/x/sync/errgroup"
@@ -95,14 +97,6 @@ type streamIngestionProcessor struct {
 	// and have attempted to flush them with `internalDrained`.
 	internalDrained bool
 
-	// ingestionErr stores any error that is returned from the worker goroutine so
-	// that it can be forwarded through the DistSQL flow.
-	ingestionErr error
-
-	// pollingErr stores any error that is returned from the poller checking for a
-	// cutover signal so that it can be forwarded through the DistSQL flow.
-	pollingErr error
-
 	// pollingWaitGroup registers the polling goroutine and waits for it to return
 	// when the processor is being drained.
 	pollingWaitGroup sync.WaitGroup
@@ -117,6 +111,20 @@ type streamIngestionProcessor struct {
 	// closePoller is used to shutdown the poller that checks the job for a
 	// cutover signal.
 	closePoller chan struct{}
+
+	// mu is used to provide thread-safe read-write operations to ingestionErr
+	// and pollingErr.
+	mu struct {
+		syncutil.Mutex
+
+		// ingestionErr stores any error that is returned from the worker goroutine so
+		// that it can be forwarded through the DistSQL flow.
+		ingestionErr error
+
+		// pollingErr stores any error that is returned from the poller checking for a
+		// cutover signal so that it can be forwarded through the DistSQL flow.
+		pollingErr error
+	}
 }
 
 // partitionEvent augments a normal event with the partition it came from.
@@ -190,7 +198,9 @@ func (sip *streamIngestionProcessor) Start(ctx context.Context) {
 		defer sip.pollingWaitGroup.Done()
 		err := sip.checkForCutoverSignal(ctx, sip.closePoller)
 		if err != nil {
-			sip.pollingErr = errors.Wrap(err, "error while polling job for cutover signal")
+			sip.mu.Lock()
+			sip.mu.pollingErr = errors.Wrap(err, "error while polling job for cutover signal")
+			sip.mu.Unlock()
 		}
 	}()
 
@@ -220,8 +230,11 @@ func (sip *streamIngestionProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.Pr
 		return nil, sip.DrainHelper()
 	}
 
-	if sip.pollingErr != nil {
-		sip.MoveToDraining(sip.pollingErr)
+	sip.mu.Lock()
+	err := sip.mu.pollingErr
+	sip.mu.Unlock()
+	if err != nil {
+		sip.MoveToDraining(err)
 		return nil, sip.DrainHelper()
 	}
 
@@ -243,8 +256,11 @@ func (sip *streamIngestionProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.Pr
 		return row, nil
 	}
 
-	if sip.ingestionErr != nil {
-		sip.MoveToDraining(sip.ingestionErr)
+	sip.mu.Lock()
+	err = sip.mu.ingestionErr
+	sip.mu.Unlock()
+	if err != nil {
+		sip.MoveToDraining(err)
 		return nil, sip.DrainHelper()
 	}
 
@@ -372,7 +388,10 @@ func (sip *streamIngestionProcessor) merge(
 		})
 	}
 	go func() {
-		sip.ingestionErr = g.Wait()
+		err := g.Wait()
+		sip.mu.Lock()
+		defer sip.mu.Unlock()
+		sip.mu.ingestionErr = err
 		close(merged)
 	}()
 
@@ -405,6 +424,14 @@ func (sip *streamIngestionProcessor) consumeEvents() (*jobspb.ResolvedSpans, err
 				return sip.flush()
 			}
 
+			if streamingKnobs, ok := sip.FlowCtx.TestingKnobs().StreamingTestingKnobs.(*sql.StreamingTestingKnobs); ok {
+				if streamingKnobs != nil {
+					if streamingKnobs.RunAfterReceivingEvent != nil {
+						streamingKnobs.RunAfterReceivingEvent(sip.Ctx)
+					}
+				}
+			}
+
 			switch event.Type() {
 			case streamingccl.KVEvent:
 				if err := sip.bufferKV(event); err != nil {
@@ -426,6 +453,15 @@ func (sip *streamIngestionProcessor) consumeEvents() (*jobspb.ResolvedSpans, err
 				}
 
 				return sip.flush()
+			case streamingccl.GenerationEvent:
+				log.Info(sip.Ctx, "GenerationEvent received")
+				select {
+				case <-sip.cutoverCh:
+					sip.internalDrained = true
+					return nil, nil
+				case <-sip.Ctx.Done():
+					return nil, sip.Ctx.Err()
+				}
 			default:
 				return nil, errors.Newf("unknown streaming event type %v", event.Type())
 			}

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_test.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_test.go
@@ -12,6 +12,7 @@ import (
 	"context"
 	"fmt"
 	"strconv"
+	"sync"
 	"testing"
 	"time"
 
@@ -24,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
@@ -70,7 +72,6 @@ func (m *mockStreamClient) ConsumePartition(
 	}
 
 	eventCh := make(chan streamingccl.Event, len(events))
-
 	for _, event := range events {
 		eventCh <- event
 	}
@@ -170,6 +171,58 @@ func TestStreamIngestionProcessor(t *testing.T) {
 		row, meta := out.Next()
 		require.Nil(t, row)
 		testutils.IsError(meta.Err, "this client always returns an error")
+	})
+
+	t.Run("stream ingestion processor shuts down gracefully on losing client connection", func(t *testing.T) {
+		events := []streamingccl.Event{streamingccl.MakeGenerationEvent()}
+		pa := streamingccl.PartitionAddress("partition")
+		mockClient := &mockStreamClient{
+			partitionEvents: map[streamingccl.PartitionAddress][]streamingccl.Event{pa: events},
+		}
+
+		startTime := hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+		partitionAddresses := []streamingccl.PartitionAddress{"partition"}
+
+		processEventCh := make(chan struct{})
+		defer close(processEventCh)
+		streamingTestingKnob := &sql.StreamingTestingKnobs{RunAfterReceivingEvent: func(ctx context.Context) {
+			processEventCh <- struct{}{}
+		}}
+		sip, out, err := getStreamIngestionProcessor(ctx, t, registry, kvDB, "randomgen://test/",
+			partitionAddresses, startTime, nil /* interceptEvents */, mockClient, streamingTestingKnob)
+		require.NoError(t, err)
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			sip.Run(ctx)
+		}()
+
+		// The channel will block on read if the event has not been intercepted yet.
+		// Once it unblocks, we are guaranteed that the mockClient has sent the
+		// GenerationEvent and the processor has read it.
+		<-processEventCh
+
+		// The sip processor has received a GenerationEvent and is thus
+		// waiting for a cutover signal, so let's send one!
+		sip.cutoverCh <- struct{}{}
+
+		wg.Wait()
+		// Ensure that all the outputs are properly closed.
+		if !out.ProducerClosed() {
+			t.Fatalf("output RowReceiver not closed")
+		}
+
+		for {
+			// No metadata should have been produced since the processor
+			// should have been moved to draining state with a nil error.
+			row := out.NextNoMeta(t)
+			if row == nil {
+				break
+			}
+			t.Fatalf("more output rows than expected")
+		}
 	})
 }
 
@@ -379,6 +432,31 @@ func runStreamIngestionProcessor(
 	interceptEvents []streamclient.InterceptFn,
 	mockClient streamclient.Client,
 ) (*distsqlutils.RowBuffer, error) {
+	sip, out, err := getStreamIngestionProcessor(ctx, t, registry, kvDB, streamAddr,
+		partitionAddresses, startTime, interceptEvents, mockClient, nil /* streamingTestingKnobs */)
+	require.NoError(t, err)
+
+	sip.Run(ctx)
+
+	// Ensure that all the outputs are properly closed.
+	if !out.ProducerClosed() {
+		t.Fatalf("output RowReceiver not closed")
+	}
+	return out, err
+}
+
+func getStreamIngestionProcessor(
+	ctx context.Context,
+	t *testing.T,
+	registry *jobs.Registry,
+	kvDB *kv.DB,
+	streamAddr string,
+	partitionAddresses []streamingccl.PartitionAddress,
+	startTime hlc.Timestamp,
+	interceptEvents []streamclient.InterceptFn,
+	mockClient streamclient.Client,
+	streamingTestingKnobs *sql.StreamingTestingKnobs,
+) (*streamIngestionProcessor, *distsqlutils.RowBuffer, error) {
 	st := cluster.MakeTestingClusterSettings()
 	evalCtx := tree.MakeTestingEvalContext(st)
 
@@ -387,9 +465,10 @@ func runStreamIngestionProcessor(
 
 	flowCtx := execinfra.FlowCtx{
 		Cfg: &execinfra.ServerConfig{
-			Settings:    st,
-			DB:          kvDB,
-			JobRegistry: registry,
+			Settings:     st,
+			DB:           kvDB,
+			JobRegistry:  registry,
+			TestingKnobs: execinfra.TestingKnobs{StreamingTestingKnobs: streamingTestingKnobs},
 		},
 		EvalCtx:     &evalCtx,
 		DiskMonitor: testDiskMonitor,
@@ -423,14 +502,7 @@ func runStreamIngestionProcessor(
 			interceptable.RegisterInterception(interceptor)
 		}
 	}
-
-	sip.Run(ctx)
-
-	// Ensure that all the outputs are properly closed.
-	if !out.ProducerClosed() {
-		t.Fatalf("output RowReceiver not closed")
-	}
-	return out, err
+	return sip, out, err
 }
 
 func registerValidatorWithClient(

--- a/pkg/ccl/streamingccl/streamingtest/replication_helpers.go
+++ b/pkg/ccl/streamingccl/streamingtest/replication_helpers.go
@@ -53,6 +53,14 @@ func ResolvedAtLeast(lo hlc.Timestamp) FeedPredicate {
 	}
 }
 
+// ReceivedNewGeneration makes a FeedPredicate that matches when a GenerationEvent has
+// been received.
+func ReceivedNewGeneration() FeedPredicate {
+	return func(msg streamingccl.Event) bool {
+		return msg.Type() == streamingccl.GenerationEvent
+	}
+}
+
 // FeedSource is a source of events for a ReplicationFeed.
 type FeedSource interface {
 	// Next returns the next event, and a flag indicating if there are more events
@@ -90,6 +98,12 @@ func (rf *ReplicationFeed) ObserveKey(ctx context.Context, key roachpb.Key) roac
 func (rf *ReplicationFeed) ObserveResolved(ctx context.Context, lo hlc.Timestamp) hlc.Timestamp {
 	require.NoError(rf.t, rf.consumeUntil(ctx, ResolvedAtLeast(lo)))
 	return *rf.msg.GetResolved()
+}
+
+// ObserveGeneration consumes the feed until we received a GenerationEvent. Returns true.
+func (rf *ReplicationFeed) ObserveGeneration(ctx context.Context) bool {
+	require.NoError(rf.t, rf.consumeUntil(ctx, ReceivedNewGeneration()))
+	return true
 }
 
 // Close cleans up any resources.

--- a/pkg/sql/exec_util.go
+++ b/pkg/sql/exec_util.go
@@ -990,6 +990,7 @@ type ExecutorConfig struct {
 	EvalContextTestingKnobs       tree.EvalContextTestingKnobs
 	TenantTestingKnobs            *TenantTestingKnobs
 	BackupRestoreTestingKnobs     *BackupRestoreTestingKnobs
+	StreamingTestingKnobs         *StreamingTestingKnobs
 	IndexUsageStatsTestingKnobs   *idxusage.TestingKnobs
 	SQLStatsTestingKnobs          *persistedsqlstats.TestingKnobs
 	// HistogramWindowInterval is (server.Config).HistogramWindowInterval.
@@ -1265,6 +1266,18 @@ var _ base.ModuleTestingKnobs = &BackupRestoreTestingKnobs{}
 
 // ModuleTestingKnobs implements the base.ModuleTestingKnobs interface.
 func (*BackupRestoreTestingKnobs) ModuleTestingKnobs() {}
+
+// StreamingTestingKnobs contains knobs for streaming behavior.
+type StreamingTestingKnobs struct {
+	// RunAfterReceivingEvent allows blocking the stream ingestion processor after
+	// a single event has been received.
+	RunAfterReceivingEvent func(ctx context.Context)
+}
+
+var _ base.ModuleTestingKnobs = &StreamingTestingKnobs{}
+
+// ModuleTestingKnobs implements the base.ModuleTestingKnobs interface.
+func (*StreamingTestingKnobs) ModuleTestingKnobs() {}
 
 func shouldDistributeGivenRecAndMode(
 	rec distRecommendation, mode sessiondatapb.DistSQLExecMode,

--- a/pkg/sql/execinfra/server_config.go
+++ b/pkg/sql/execinfra/server_config.go
@@ -243,6 +243,9 @@ type TestingKnobs struct {
 
 	// BackupRestoreTestingKnobs are backup and restore specific testing knobs.
 	BackupRestoreTestingKnobs base.ModuleTestingKnobs
+
+	// StreamingTestingKnobs are backup and restore specific testing knobs.
+	StreamingTestingKnobs base.ModuleTestingKnobs
 }
 
 // MetadataTestLevel represents the types of queries where metadata test

